copier: init at 7.0.1

jonringer opened the pull request:
###### Motivation for this change
Thought I would try my hand at packaging this, someone mentioned they had great difficulty doing so, and I would agree.

https://discourse.nixos.org/t/buildpythonpackage-throwing-dependency-version-errors-when-i-dont-belive-it-should-be/15818

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
python310Packages.iteration-utilities: init at 0.11.0

Commit:
python310Packages.jinja2-ansible-filters: init at 1.3.2

Commit:
python310Packages.mkdocs-mermaid2-plugin: init at 0.6.0

Commit:
python310Packages.pyyaml-include: init at 1.3

Commit:
copier: init at 7.0.1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/buildpythonpackage-throwing-dependency-version-errors-when-i-dont-belive-it-should-be/15818/2

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/buildpythonpackage-throwing-dependency-version-errors-when-i-dont-belive-it-should-be/15818/3

fabaff commented on pkgs/tools/misc/copier/default.nix:
```diff
@@ -0,0 +1,61 @@
+  '';
+
+  meta = with lib; {
+    description = "Library and command-line utility for rendering projects templates.";
```
```suggestion
    description = "Library and command-line utility for rendering projects templates";
```

fabaff commented on pkgs/tools/misc/copier/default.nix:
```diff
@@ -0,0 +1,61 @@
+  '';
+
+  nativeBuildInputs = [
+    poetry
```
Upstream seems to use `poetry-core` as [build backend](https://github.com/copier-org/copier/blob/master/pyproject.toml#L120).

fabaff commented on pkgs/tools/misc/copier/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  meta = with lib; {
+    description = "Library and command-line utility for rendering projects templates.";
+    homepage = "github.com/copier-org/copier";
```
Opinionated change

```suggestion
    homepage = "https://copier.readthedocs.io";
```

fabaff commented on pkgs/development/python-modules/dunamai/default.nix:
```diff
@@ -0,0 +1,50 @@
+  };
+
+  nativeBuildInputs = [
+    poetry
```
Perhaps https://github.com/mtkennerly/dunamai/pull/35 gets merged quickly then we could move to `poetry-core`.

fabaff commented on pkgs/development/python-modules/jinja2-ansible-filters/default.nix:
```diff
@@ -0,0 +1,36 @@
+  ];
+
+  # no tests include in sdist, and source not available
+  doCheck = false;
```
The upstream repo contains a couple of [`tests`](https://gitlab.com/dreamer-labs/libraries/jinja2-ansible-filters/-/tree/master/tests).

fabaff commented on pkgs/development/python-modules/poetry-dynamic-versioning/default.nix:
```diff
@@ -0,0 +1,30 @@
+    jinja2
+    tomlkit
+  ];
+
```
[Tests](https://github.com/mtkennerly/poetry-dynamic-versioning/tree/master/tests) would be available if the source from GitHub is used.

SuperSandro2000 commented on pkgs/tools/misc/copier/default.nix:
```diff
@@ -0,0 +1,61 @@
+  };
+
+  postPatch = ''
+    ${gnused}/bin/sed -e '/packaging/d' -i pyproject.toml
```
```suggestion
    sed -e '/packaging/d' -i pyproject.toml
```

SuperSandro2000 commented on pkgs/tools/misc/copier/default.nix:
```diff
@@ -0,0 +1,61 @@
+{ lib, gnused, poetry, git, python3, fetchFromGitHub }:
```
```suggestion
{ lib, poetry, git, python3, fetchFromGitHub }:
```
sed is part of the stdenv.

SuperSandro2000 commented on pkgs/development/python-modules/pyyaml-include/default.nix:
```diff
@@ -0,0 +1,37 @@
+    pyyaml
+    setuptools-scm
+    setuptools-scm-git-archive
+    toml
```
That should be part of setuptools-scm actually.

SuperSandro2000 commented on pkgs/development/python-modules/plumbum/default.nix:
```diff
@@ -1,19 +1,64 @@
+    description = "Shell combinators";
+    homepage = "https://github.com/tomerfiliba/plumbum";
+    license = licenses.mit;
+    maintainers = [ ];
```
The contributing guide wants us to set this.

SuperSandro2000 commented on pkgs/development/python-modules/plumbum/default.nix:
```diff
@@ -1,19 +1,64 @@
+  checkInputs = [
+    paramiko
+    psutil
+    pytest-cov
```
```suggestion
```
collecting coverage data has no benefit for us or upstream except that it takes ressources.

samuelludwig commented:
Unfortunately it seems the main functionality of this package is broken, in a similar way as to when I made my attempt.

`${copier} --help` prints whats expected correctly, however, beyond just the help message, little works as it ought to:

If you set up a simple template like so:

```bash
mkdir my-template
cd my-template
echo "filename:
  type: str
  help: What's the name of the file?
  default: myDefaultName
  " > copier.yml
echo "Hi, I'm [[ filename ]]" > '[[ filename ]].txt.tmpl'
```

and then run `copier my-template cop-test`

you should get a prompt requesting a value for `filename`, supposing we entered "working", in `cop-test` there should then be one file named `working.txt` with the contents "`Hi, I'm working`".

When _this_ package is used, versus the copier installed via `pipx`, a similar (but not identical) prompt appears, and seems to accept a value, however you will find that only the raw file is copied over, with no substitutions done. So in `cop-test` there will be a file named `'[[ filename ]].txt.tmpl'` with the contents "`Hi, I'm [[ filename ]]`".

jonringer commented:
> When this package is used, versus the copier installed via pipx, a similar (but not identical) prompt appears, and seems to accept a value, however you will find that only the raw file is copied over, with no substitutions done. So in cop-test there will be a file named '[[ filename ]].txt.tmpl' with the contents "Hi, I'm [[ filename ]]".

This seems to be intended behavior https://github.com/copier-org/copier/blob/master/CHANGELOG.md#changed 
> We use Jinja 2 defaults now. {{ }} instead of [[ ]] and similar.

One thing to note is that I'm packaging off their master since it's more coherent with the rest of python packages. Breaking changes won't be reflected.

samuelludwig commented:
> 
> 
> > When this package is used, versus the copier installed via pipx, a similar (but not identical) prompt appears, and seems to accept a value, however you will find that only the raw file is copied over, with no substitutions done. So in cop-test there will be a file named '[[ filename ]].txt.tmpl' with the contents "Hi, I'm [[ filename ]]".
> 
> This seems to be intended behavior https://github.com/copier-org/copier/blob/master/CHANGELOG.md#changed
> 
> > We use Jinja 2 defaults now. {{ }} instead of [[ ]] and similar.
> 
> One thing to note is that I'm packaging off their master since it's more coherent with the rest of python packages. Breaking changes won't be reflected.

oh no... that means i might have had it in an actual working state before... my apologies, thank you, I at least for sure didn't properly package any of the dependencies for nix as you did here

fbergroth commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4792,6 +4792,8 @@ in {
 
   mkdocs-material-extensions = callPackage ../development/python-modules/mkdocs-material-extensions { };
 
+  mkdocs-mermaid2-plugin = callPackage ../development/python-modules/mkdocs2-mermaid2-plugin { };
```
`mkdocs2` in `mkdocs2-mermaid2-plugin` looks like a typo

hhoeflin commented:
Would be great if this could be merged! Thanks

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/707

jonringer commented:
related issue: https://github.com/mtkennerly/poetry-dynamic-versioning/issues/62

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/help-building-pdm-based-python-package/18256/2

yajo commented on pkgs/tools/misc/copier/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+python3.pkgs.buildPythonApplication rec {
+  pname = "copier";
+  version = "unstable-2021-11-10";
```
FWIW [Copier 6 was recently released](https://github.com/copier-org/copier/releases/tag/v6.0.0). You'd probably prefer to package that?

Artturin committed:
python310Packages.poetry-dynamic-versioning: disable tests

 E           RuntimeError: The command 'poetry build' returned code 2. Output:
 E           Creating virtualenv poetry-dynamic-versioning-nj6DJBl3-py3.10 in /homeless-shelter/.cache/pypoetry/virtualenvs
 E           usage: virtualenv ...
 E           virtualenv: error: argument dest: the destination . is not write-able at /
 tests/test_integration.py:43: RuntimeError
